### PR TITLE
Add Tensor overload for start in narrow.

### PR DESCRIPTION
**Pull request by ailzhang**
#31558

### CODE REVIEW

File: aten/src/ATen/native/TensorShape.cpp [hunk 1]
```diff
@@ -587,6 +587,14 @@ Tensor narrow(const Tensor& self, int64_t dim, int64_t start, int64_t length) {
   return at::slice(self, dim, start, start + length, 1);
 }
 
+Tensor narrow(const Tensor& self, int64_t dim, const Tensor& start, int64_t length) {
+  TORCH_CHECK(self.dim() > 0, "narrow() cannot be applied to a 0-dim tensor.");
```
Comment (gchanan) [inline aten/src/ATen/native/TensorShape.cpp]: can we just do this check in the other narrow function?

File: aten/src/ATen/native/TensorShape.cpp [hunk 2]
```diff
@@ -587,6 +587,14 @@ Tensor narrow(const Tensor& self, int64_t dim, int64_t start, int64_t length) {
...
+  TORCH_CHECK(start.numel() == 1 && isIntegralType(start.scalar_type(), /*includeBool=*/false),
```
Comment (gchanan) [inline aten/src/ATen/native/TensorShape.cpp]: make the error message the same as the thing you check -- i.e. "1-element integral tensor".

You can probably not mention "int" because that will be checked in the other code path.

Comment (gchanan) [reply aten/src/ATen/native/TensorShape.cpp]: does the current python index handling allow any integer type?

Comment (ailzhang) [reply aten/src/ATen/native/TensorShape.cpp]: Updated: current python index handling allow any integer type 0-dim tensor.

File: test/test_torch.py [hunk 3]
```diff
@@ -2047,6 +2047,16 @@ def test_narrow(self):
         self.assertEqual(x.narrow(-1, -1, 1), torch.Tensor([[2], [5], [8]]))
         self.assertEqual(x.narrow(-2, -1, 1), torch.Tensor([[6, 7, 8]]))
 
+    def test_narrow_tensor(self):
+        x = torch.Tensor([[0, 1, 2], [3, 4, 5], [6, 7, 8]])
+        self.assertEqual(x.narrow(0, torch.tensor(0), 1), torch.Tensor([[0, 1, 2]]))
+        with self.assertRaises(Exception):
+            x.narrow(0, torch.tensor(0, dtype=torch.float), 1)
```
Comment (gchanan) [inline test/test_torch.py]: nit: you can do this more succinctly with `torch.tensor(0.)`.

**Review by gchanan — APPROVED** [review]
